### PR TITLE
Add 'attacking rate' option to the attacking tool.

### PR DESCRIPTION
**Pull request by VipAlekseyPetrenko**
## Description

Enable the attacking tool to control the attacking rate, to allow for granular comparison between different DoS attack types.

## Testing

Manually tested my changes with an attacker VM that has XDP option enabled and a target VM that has Windows HTTP server running.

## Documentation

n/a

### CODE REVIEW

File: src/tools/attack/attack.cpp [hunk 1]
```diff
@@ -22,7 +22,9 @@
 
 #define ATTACK_TIMEOUT_DEFAULT_MS (60 * 1000)
 
-#define ATTACK_THREADS_DEFAULT CxPlatProcCount()
+#define ATTACK_THREADS_DEFAULT max(1, CxPlatProcCount())
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:25]: This should be unnecessary. Processor count will never be zero

File: src/tools/attack/attack.cpp [hunk 2]
```diff
@@ -152,21 +157,34 @@ void RunAttackRandom(CXPLAT_SOCKET* Binding, uint16_t Length, bool ValidQuic, bo
...
     CxPlatRandom(sizeof(ConnectionId), &ConnectionId);
 
-    while (CxPlatTimeDiff64(TimeStart, CxPlatTimeMs64()) < TimeoutMs) {
+    uint64_t BucketTime = CxPlatTimeMs64(), CurTime;
+    uint64_t BucketCount = 0;
+    uint64_t BucketThreshold = max(1, AttackRate / ThreadCount);
+    
+    while (CxPlatTimeDiff64(TimeStart, (CurTime = CxPlatTimeMs64())) < TimeoutMs) {
+
+        if(CxPlatTimeDiff64(BucketTime, CurTime) > 1000) {
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:166]:
```suggestion
        if (CxPlatTimeDiff64(BucketTime, CurTime) > 1000) {
```

File: src/tools/attack/attack.cpp [hunk 3]
```diff
@@ -252,7 +272,19 @@ void RunAttackValidInitial(CXPLAT_SOCKET* Binding)
...
+    uint64_t BucketCount = 0;
+    uint64_t BucketThreshold = max(1, AttackRate / ThreadCount);;
+
+    while (CxPlatTimeDiff64(TimeStart, (CurTime = CxPlatTimeMs64())) < TimeoutMs) {
+
+        if(CxPlatTimeDiff64(BucketTime, CurTime) > 1000) {
+            BucketTime = CurTime;
+            BucketCount = 0;
+        }
+        if (BucketCount >= BucketThreshold) {
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:285]:
```suggestion
        }

        if (BucketCount >= BucketThreshold) {
```

File: submodules/googletest
Comment (nibanks) [inline submodules/googletest:1]: Please undo this change

Comment (VipAlekseyPetrenko) [reply submodules/googletest:1]: Done.

File: src/tools/attack/attack.cpp (hunk 1 above)
Comment (VipAlekseyPetrenko) [reply src/tools/attack/attack.cpp]: Good point.

File: src/tools/attack/attack.cpp [hunk 4]
```diff
@@ -50,18 +53,18 @@ void PrintUsage()
...
     printf("  quicattack.exe -list\n\n");
-    printf("  quicattack.exe -type:<number> -ip:<ip_address_and_port> [-alpn:<protocol_name>] [-sni:<host_name>] [-timeout:<ms>] [-threads:<count>]\n\n");
+    printf("  quicattack.exe -type:<number> -ip:<ip_address_and_port> [-alpn:<protocol_name>] [-sni:<host_name>] [-timeout:<ms>] [-threads:<count>] [-rate:<packet_rate>]\n\n");
 }
 
 void PrintUsageList()
 {
     printf("The following are the different types of attacks supported by the tool.\n\n");
-
+
```
Comment (nibanks) [inline src/tools/attack/attack.cpp:62]: Please remove these spaces.

File: submodules/googletest
Comment (nibanks) [inline submodules/googletest:1]: You still have this unnecessary change.

**Review by nibanks — APPROVED** [review]

**Review by nibanks — APPROVED** [review]
